### PR TITLE
Handle dragging improvements

### LINKED ISSUE #110305: Cupertino style text field and Native iOS UITextField has different behavior when move between different lines
Flutter CupertinoTextField：

https://user-images.githubusercontent.com/30322203/185344552-0239b156-ecd3-437c-93db-000df1d2cc60.mov

iOS UITextField:

https://user-images.githubusercontent.com/30322203/186804029-9ec8cd6e-8185-4d22-bf5f-c33daaf2912a.mov

 can reproduce in any flutter cupertino textfield

### PR DESCRIPTION
**Pull request by justinmc**
Recently in https://github.com/flutter/flutter/pull/109136, a bug was fixed where text selection handles would jump around while being dragged.  However, investigation in that PR revealed that the behavior when moving between lines was still incorrect.  This behavior fixes the behavior when moving between lines.

The correct behavior is that the handle jumps between lines at the point where the cursor would be touching the handle on the new line at the point where the drag began.  Below, notice that in the "after" gifs, the cursor is roughly on the same place on the handle when the line changes.

| Platform | Before | After |
| --- | --- | --- |
| iOS |output | output |
| Android | incorrect_android | correct_android |

Fixes https://github.com/flutter/flutter/issues/110305

### CODE REVIEW

**Comment by justinmc** [comment]
FYI @ksballetba this is my attempt at fixing the problem of when to trigger a vertical line change, discussed in your PR https://github.com/flutter/flutter/pull/109136.

File: packages/flutter/lib/src/widgets/text_selection.dart [hunk 1]
```diff
@@ -664,10 +701,15 @@ class TextSelectionOverlay {
...
-    final Offset offsetFromHandleToTextPosition = _getOffsetToTextPositionPoint(_selectionOverlay.startHandleType);
-    _dragStartPosition = details.globalPosition + offsetFromHandleToTextPosition;
-
-    final TextPosition position = renderObject.getPositionForPoint(_dragStartPosition);
+    _startHandleDragPosition = details.globalPosition.dy;
+    final Offset startPoint =
+        renderObject.localToGlobal(_selectionOverlay.selectionEndpoints.first.point);
+    final double centerOfLine = startPoint.dy - renderObject.preferredLineHeight / 2;
+    _startHandleDragPositionToCenterOfLine = centerOfLine - _startHandleDragPosition;
+    final TextPosition position = renderObject.getPositionForPoint(Offset(
```
Comment (Renzo-Olivares) [inline packages/flutter/lib/src/widgets/text_selection.dart:709]:
```dart
final TextPosition position = renderObject.getPositionForPoint(
	Offset(
	  details.globalPosition.dx,
	  centerOfLine,
	),
);
```

Comment (Renzo-Olivares) [inline packages/flutter/lib/src/widgets/text_selection.dart:709]: Nit: formatting maybe

```dart
final TextPosition position = renderObject.getPositionForPoint(
  Offset(
    details.globalPosition.dx,
    centerOfLine,
  ),
);
```

File: packages/flutter/lib/src/widgets/text_selection.dart [hunk 2]
```diff
@@ -585,10 +591,15 @@ class TextSelectionOverlay {
...
-    final Offset offsetFromHandleToTextPosition = _getOffsetToTextPositionPoint(_selectionOverlay.endHandleType);
-    _dragEndPosition = details.globalPosition + offsetFromHandleToTextPosition;
-
-    final TextPosition position = renderObject.getPositionForPoint(_dragEndPosition);
+    _endHandleDragPosition = details.globalPosition.dy;
+    final Offset endPoint =
+        renderObject.localToGlobal(_selectionOverlay.selectionEndpoints.last.point);
+    final double centerOfLine = endPoint.dy - renderObject.preferredLineHeight / 2;
+    _endHandleDragPositionToCenterOfLine = centerOfLine - _endHandleDragPosition;
+    final TextPosition position = renderObject.getPositionForPoint(Offset(
```
Comment (Renzo-Olivares) [inline packages/flutter/lib/src/widgets/text_selection.dart:599]: Nit: formatting, see similar comment below.

**Review by Renzo-Olivares — APPROVED** [review]
Wow awesome change! That looks much better!

LGTM

**Comment by justinmc** [comment]
The Google test failure seems like a flake or something...  It was a generic "flutter run" test that failed to start the app.  I'm merging master in so it will run again.

**Comment by justinmc** [comment]
I'll plan to manually set the Google Testing to pass after discussing in hackers-infra on Discord.  Seems to be a flake.

File: packages/flutter/test/material/text_field_test.dart [hunk 3]
```diff
@@ -2537,6 +2537,381 @@ void main() {
...
+    // Drag the left handle 1 line up.
+    endpoints = globalize(
+      renderEditable.getEndpointsForSelection(controller.selection),
+      renderEditable,
+    );
+    handlePos = endpoints[0].point + startHandleAdjustment;
+    newHandlePos = handlePos - toNextLine;
+    gesture = await tester.startGesture(handlePos, pointer: 7);
+    await tester.pump();
+    await gesture.moveTo(newHandlePos);
```
Comment (Renzo-Olivares) [inline packages/flutter/test/material/text_field_test.dart:2877]: Referring to this test.

**Pull request merged**